Framework/Algorithms/src/EstimateResolutionDiffraction.cpp

```diff
@@ -225,100 +226,130 @@ void EstimateResolutionDiffraction::estimateDetectorResolution() {
+                                    const double centraltof = (l1 + l2) / centreVel;
+                                    const double term = deltaT / centraltof;
+                                    return sum + (term * term);
+                                  }
+                                });
+      term1Sq = term1Sq / numDet;
     }
 
     // resolution in length
-    double detdimSq;
-    const auto realdet = dynamic_cast<const Detector *>(&det);
-    if (realdet) {
-      const double dy = realdet->getHeight();
-      const double dx = realdet->getWidth();
-      detdimSq = (dx * dx + dy * dy) * (0.5 * 0.5);
-    } else {
-      // Use detector dimension as 0 as no-information
-      detdimSq = 0;
-      ++count_nodetsize;
+    double term2Sq;
+    { // reduce scope
+      const double sourceTerm = m_sourceDeltaLMetersSq;
+      term2Sq = std::accumulate(spectrumDefinition.cbegin(), spectrumDefinition.cend(), 0.,
+                                [&detectorInfo, &sourceTerm, &l1](const auto sum, const auto &index) {
+                                  if (detectorInfo.isMasked(index.first) || detectorInfo.isMonitor(index.first)) {
+                                    return sum;
+                                  } else {
+                                    const auto &detector = detectorInfo.detector(index.first);
+                                    const auto realdet = dynamic_cast<const Detector *>(&detector);
+                                    if (realdet) {
+                                      const auto l2 = detectorInfo.l2(index.first);
+                                      const auto l_total = (l1 + l2);
+                                      realdet->shape();
+                                      const double dx = realdet->getWidth();
+                                      const double dy = realdet->getHeight();
+                                      const double dz = realdet->getDepth();
+                                      const double detdimSq = (dx * dx + dy * dy + dz * dz) * (0.5 * 0.5);
+                                      const double term = (detdimSq + sourceTerm) / (l_total * l_total);
+                                      return sum + term;
+                                    } else {
+                                      return sum;
+                                    }
+                                  }
+                                });
+      term2Sq = term2Sq / numDet;
     }
-    const double l_total = (l1 + l2);
-    const double term2Sq = (detdimSq + m_sourceDeltaLMetersSq) / (l_total * l_total);
 
-    // resolution in angle
-    const double twotheta = spectrumInfo.isMonitor(i) ? 0.0 : spectrumInfo.twoTheta(i);
-    const double theta = 0.5 * twotheta;
+    // resolution in angle - everything is in radians
+    const double theta = spectrumInfo.isMonitor(specNum) ? 0.0 : 0.5 * spectrumInfo.twoTheta(specNum);
```

Comment:
why not maintain `twotheta`?  It avoids having to write `2.0 * theta` twice below